cmd/lakectl/cmd/annotate.go

```diff
@@ -0,0 +1,84 @@
+package cmd
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/spf13/cobra"
+	"github.com/treeverse/lakefs/pkg/api"
+)
+
+const annotateTemplate = `{{  $val := .Commit}}
+{{ $.Object|ljust 15}} {{ $val.Committer|ljust 20 }} {{ $val.Id | printf "%.16s"|ljust 20 }} {{ $val.CreationDate|date }}  {{ $.CommitMessage |ljust 30 }}
+`
+const messageSize = 100
+const trailingThreeDots = "..."
+
+var annotateCmd = &cobra.Command{
+	Use:     "annotate <path uri>",
+	Short:   "List entries under a given path, annotating each with the latest modifying commit ",
+	Aliases: []string{"blame"},
+	Args:    cobra.ExactArgs(1),
+	Run: func(cmd *cobra.Command, args []string) {
+		pathURI := MustParsePathURI("path", args[0])
+		client := getClient()
+		amount := 1
+		pfx := api.PaginationPrefix(*pathURI.Path)
+		context := cmd.Context()
+		res, err := client.ListObjectsWithResponse(context, pathURI.Repository, pathURI.Ref, &api.ListObjectsParams{Prefix: &pfx})
+		DieOnResponseError(res, err)
+		var from string
+		for {
+			params := &api.ListObjectsParams{
+				Prefix: &pfx,
+				After:  api.PaginationAfterPtr(from),
+			}
+			resp, err := client.ListObjectsWithResponse(context, pathURI.Repository, pathURI.Ref, params)
+			DieOnResponseError(resp, err)
+			for _, obj := range resp.JSON200.Results {
+				path := []string{obj.Path}
+				logCommitsParams := &api.LogCommitsParams{
+					Amount:  api.PaginationAmountPtr(amount),
+					Objects: &path,
```

Comment:
For a large prefix this can be quite inefficient.  I believe that in many cases something like the following will be (much) faster:

* For each 1000 objects, put them in a set, and starting from HEAD commit...
   * Take the _latest_ earlier commit that mentions _any_ object(s) in that set.  [Single API call]
   * Link each of those objects in the set that occur on that commit to that commit.  [Just local processing]
   * Remove the linked objects from the set
   * If set is non-empty, go back to fetching the latest commit.
* Continue with next 1000 objects.

This minimizes API calls by several 10x or 100x in most cases and _never_ increases the number of API calls.

Reply:
This is an interesting approach. I would consider implementing it if we indeed see that the inefficiency is noticable. After all, this is happing on the client end.